packages/core/core/src/factories/apiClientFactory/index.ts

```diff
@@ -21,29 +21,28 @@ const apiClientFactory = <ALL_SETTINGS extends ApiClientConfig, ALL_FUNCTIONS>(f
 
     const _config = lifecycles
       .filter(ext => isFn(ext.beforeCreate))
-      .reduce((prev, curr) => curr.beforeCreate({ configuration: prev }), config);
+      .reduce((configuration, curr) => curr.beforeCreate({ configuration }), config);
 
     const settings = factoryParams.onCreate ? factoryParams.onCreate(_config) : { config, client: config.client };
 
     Logger.debug('apiClientFactory.create', settings);
 
     settings.config = lifecycles
       .filter(ext => isFn(ext.afterCreate))
-      .reduce((prev, curr) => curr.afterCreate({ configuration: prev }), settings.config);
-
-    const extensionHooks = {
-      before: (params) => lifecycles
-        .filter(e => isFn(e.beforeCall))
-        .reduce((args, e) => e.beforeCall({ ...params, configuration: settings.config, args}), params.args),
-      after: (params) => lifecycles
-        .filter(e => isFn(e.afterCall))
-        .reduce((response, e) => e.afterCall({ ...params, configuration: settings.config, response }), params.response)
-    };
+      .reduce((configuration, curr) => curr.afterCreate({ configuration }), settings.config);
+
+    const before = async (params) => lifecycles
+      .filter(e => isFn(e.beforeCall))
+      .reduce((args, e) => args.then(args => e.beforeCall({ ...params, configuration: settings.config, args})), Promise.resolve(params.args));
+
+    const after = async (params) => lifecycles
+      .filter(e => isFn(e.afterCall))
+      .reduce((response, e) => response.then(response => e.afterCall({ ...params, configuration: settings.config, response })), Promise.resolve(params.response));
```

Comment:
please create a functions that will wrap the repetitive things:

```js
const handlePromise = (params, hookName, fieldName) => result =>
   e.[hookName]({ ...params, configuration: settings.config, [fieldName]: result  })

const createPromises =  (params, hookName, fieldName) =>  (hook, e) =>
   hook.then(handlePromise(params, hookName, fieldName)), Promise.resolve(params[fieldName]));

```

then use as:

```js
.reduce(createPromises('beforeCall', 'args'));

.reduce(createPromises('afterCall', 'response'));
```

isn't simpler?